include/wil/rpc_helpers.h

```diff
@@ -0,0 +1,170 @@
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_RPC_HELPERS_INCLUDED
+#define __WIL_RPC_HELPERS_INCLUDED
+
+#include "result.h"
+#include "resource.h"
+#include "wistd_functional.h"
+#include "wistd_type_traits.h"
+
+namespace wil
+{
+
+    /// @cond
+    namespace details
+    {
+        // This call-adapter template converts a void-returning 'wistd::invoke' into
+        // an HRESULT-returning 'wistd::invoke' that emits S_OK. It can be eliminated
+        // with 'if constexpr' when C++17 is in wide use.
+        template<typename TReturnType> struct call_adapter
+        {
+            template<typename... TArgs> static HRESULT call(TArgs&& ... args)
+            {
+                return wistd::invoke(wistd::forward<TArgs>(args)...);
+            }
+        };
+
+        template<> struct call_adapter<void>
+        {
+            template<typename... TArgs> static HRESULT call(TArgs&& ... args)
+            {
+                wistd::invoke(wistd::forward<TArgs>(args)...);
+                return S_OK;
+            }
+        };
+
+        // Some RPC exceptions are already HRESULTs. Others are in the regular Win32
+        // error space. If the incoming exception code isn't an HRESULT, wrap it.
+        constexpr static HRESULT map_rpc_exception(DWORD code)
+        {
+            return IS_ERROR(code) ? code : __HRESULT_FROM_WIN32(code);
+        }
+
+    }
+    /// @endcond
+
+    /** Invokes an RPC method, mapping structured exceptions to HRESULTs
+    Failures encountered by the RPC infrastructure (such as server crashes, authentication
+    errors, client parameter issues, etc.) are emitted by raising a structured exception from
+    within the RPC machinery. This method wraps the requested call in the usual RpcTryExcept,
+    RpcTryCatch, and RpcEndExcept sequence then maps the exceptions to HRESULTs for the usual
+    flow control machinery to use.
+
+    Many RPC methods are defined as returning HRESULT themselves, where the HRESULT indicates
+    the result of the _work_. HRESULTs returned by a successful completion of the _call_ are
+    returned as-is.
+
+    RPC methods that have a return type of 'void' are mapped to returning S_OK when the _call_
+    completes successfully.
+
+    For example, consider an RPC interface method defined in idl as:
+    ~~~
+    HRESULT GetKittenState([in, ref, string] const wchar_t* name, [out, retval] KittenState** state);
+    ~~~
+    To call this method, use:
+    ~~~
+    wil::unique_rpc_binding binding = // typically gotten elsewhere;
+    wil::unique_midl_ptr<KittenState> state;
+    HRESULT hr = wil::invoke_rpc_nothrow(GetKittenState, binding.get(), L"fluffy", wil::out_param(state));
```

Comment:
I see it would be nice to have invoke_rpc_nothrow on unique_rpc_binding
`binding.invoke_rpc_nothrow(GetKittenState, L"fluffy", wil::out_param(state));`

Reply:
I wish we had extension methods for this. 

Sadly, there's [no requirement that the binding handle be first](https://docs.microsoft.com/en-us/windows/desktop/Rpc/explicit-binding-handles). Our RPC implementation allows it anywhere.  In some RPC methods, the binding handle is even [implicit](https://docs.microsoft.com/en-us/windows/desktop/Rpc/implicit-binding-handles)